### ISSUE TITLE
Great. improvement of build and rebuild time

### ISSUE THREAD

**Issue by linegel** [meteor/meteor-feature-requests#398]
I just played around for few hours with [esbuild](https://github.com/evanw/esbuild) and WOAH that's really. impressive thing. Usage of it could really kill that. years-long hassle with building and rebuilding of the project, we all struggling with it and that's one of main disadvantages for the usage of Meteor in a big projects. 

The main problem is looks like it could be a real mess to into this. As first step of integration I see that we could use esbuild instead of current minifier (in my projects it takes around or sometimes more than 50% of time during building stage). Then we can have a look at how we could move forward with all that we have in isobuild. It would be really nice to have a response from somebody who has worked with Meteor building process previously and who could give a tip where from I could start work on PoC

**Comment by linegel** [comment]
@filipenevola any ideas about who could help us to start work on it?

**Comment by hexsprite** [comment]
@linegel yes, I think you could start by trying to integrate this as a minifier plugin, this doesn't require any changes to Meteor core.

**Comment by StorytellerCZ** [comment]
@linegel the person who knows the build system the most probably is @benjamn and @zodern.

**Comment by crapthings** [comment]
what about this one?

https://swc.rs/

https://deno.land/manual@v1.10.3/typescript/overview#how-does-it-work

```
At a high level, Deno converts TypeScript (as well as TSX and JSX) into JavaScript. It does this via a combination of the TypeScript compiler, which we build into Deno, and a Rust library called swc. When the code has been type checked and transformed, it is stored in a cache, ready for the next run without the need to convert it from its source to JavaScript again.
```

and parcel 2b3 seems use swc

**Comment by benjamn** [comment]
As @crapthings mentioned above, the [SWC](https://swc.rs/) project aims to be a drop-in replacement for Babel, which means it could potentially replace/reimplement the [`meteor-babel`](https://github.com/meteor/babel) package, with relatively few visible changes to existing Meteor applications (beyond some new complexities with `.babelrc` files, perhaps). Using SWC in this way seems especially appealing because Babel takes up so much time (50%+) during initial (cold start) `meteor` builds.

Note: I'm not actively working on Meteor anymore, and I don't speak for the current team, but I wanted to express my relative confidence in the SWC-instead-of-Babel approach, compared to a blind replacement of the whole build system (using esbuild or any other toolchain).

**Comment by jpambrun** [comment]
We wrote a minifier plugin using esbuild instead of terser. That reduced
our build time by about 50% and we even managed to remove Meteor.isServer()
blocks from client bundles.

There is definitely something to do with esbuild and/or swc.

**Comment by StorytellerCZ** [comment]
Maybe something to consider for Meteor 3 or 4.

**Comment by harryadel** [comment]
@jpambrun Can you please create a mini repo illustrating how you've managed to do it? Or even preferably a package?

**Comment by jpambrun** [comment]
> @jpambrun Can you please create a mini repo illustrating how you've managed to do it? Or even preferably a package?

I am seeking approval.. might take a while..

**Comment by harryadel** [comment]
@jpambrun Thanks a lot!

**Comment by lscoder** [comment]
I created some tests to compare terser (faster than babel) and esbuild minifying 218 files (41MB) and esbuild (1.4s) is 46x faster than terser (65s). The final bundle had almost the same size.

**Comment by crapthings** [comment]
![image](https://user-images.githubusercontent.com/1147704/121282143-9cf42000-c90b-11eb-961a-f217f8b75a49.png)


https://github.com/meteor/meteor/blob/devel/packages/minifier-js/minifier.js

```js
terser = terser || Npm.require("terser");
  esbuild = esbuild || Npm.require('esbuild');

  result.code = esbuild.transformSync(js, {
    minify: true,
  });
  result.minifier = 'esbuild';
  return;
```

try this

**Comment by crapthings** [comment]
first clone these packagess in your project/packages

- standard-minifier-js # find package.json and change package name to scope:standard-minifier-js
- minifier-js # change name to scope:minifier-js



```js
  esbuild = esbuild || Npm.require('esbuild');
  result.code = esbuild.transformSync(source, {
    minify: true,
  }).code;
  result.minifier = 'esbuild';
  return result;
```

![image](https://user-images.githubusercontent.com/1147704/121287959-25c38980-c915-11eb-99a3-d3e7bbf92706.png)

![image](https://user-images.githubusercontent.com/1147704/121287982-2f4cf180-c915-11eb-8ea7-0171777adf44.png)

after replace i don't see performance boost with our project, weird.

the total build time almost same like before

**Comment by filipenevola** [comment]
Hi @linegel and @crapthings we are looking for alternatives like this.

@renanccastro could help in this work.

The first step would be to define the scope for a PoC and the tools that we want to try first.
